chanzuckerberg/idseq-web
# Add foreign key constraints to selected tables

# Description

This add FK constraints to tables that we expect to always have parent records, so we don't get any nasty surprises, and we always delete dependent data when requested, or raise an exception. 

# Notes

* I used the `immigrant` gem to generate the migration which generates migration for all possible tables. 
* I'm not sure whether constraints on many-to-many junction tables are a good idea. I hadn't thought about it so I ignored them all. 
* I went through all the affected assocs and updated the `dependent` directive to either `destroy` or `restrict_with_exception` so the rails behavior reflects the db behavior

# Tests

* rails db:migrate in dev, see no errors
* run in staging, verify no valid rows dropped
* verify there are no violations in prod before deploying -- some db cleanup will be required
